File: Hashing/csharp/Intersection.cs
using System.Collections.Generic;
using System.Linq;

namespace HashingSolutions;

public static class Intersection
{
    public static IList<T> Solve<T>(IEnumerable<T> first, IEnumerable<T> second)
    {
        var lookup = new HashSet<T>(second);
        var added = new HashSet<T>();
        var result = new List<T>();

        foreach (var item in first)
        {
            if (lookup.Contains(item) && added.Add(item))
            {
                result.Add(item);
            }
        }

        return result;
    }
}

File: Hashing/php/intersection.php
<?php

declare(strict_types=1);

function intersection(array $first, array $second): array
{
    $lookup = [];
    foreach ($second as $item) {
        $lookup[serialize($item)] = $item;
    }

    $added = [];
    $result = [];

    foreach ($first as $item) {
        $key = serialize($item);
        if (isset($lookup[$key]) && !isset($added[$key])) {
            $result[] = $item;
            $added[$key] = true;
        }
    }

    return $result;
}
